[Model] Add support for xverse

hxer7963 opened the pull request:
Add support for xverse models.

We tested the xverse 7B/13B/65B chat models and the quantized GPTQ model locally, and vllm responses are normal. 

You can verify by downloading xverse models from Hugging Face (https://huggingface.co/xverse) or Modelscope (https://www.modelscope.cn/search?search=xverse).

Furthermore, before the PR, we executed format.sh.

Request: After the current PR is merged, can we add a new tag so that the latest version of the package supports inference of xverse models via pip installation.

Commit:
Support xverse arch in vllm.

Commit:
Add support for xverse

Alec-Lin commented:
What is the difference of arch between llama and xverse?

hxer7963 commented:
> What is the difference of arch between llama and xverse?

Nice question.

The current xverse model architecture is no different from llama, and it is expected that xverse will add moe features within two weeks and merge them into `xverse.py`.

To maintain an independent update progress, it is necessary to separately support the xverse architecture in VLLM.

hxer7963 commented:
Hello @WoosukKwon ,

I've submitted a PR #3610 and I would greatly appreciate your help with reviewing the code changes  and provide your feedback.

Thank you for your time and assistance.

Best regards,
willhe.

WoosukKwon commented:
@hxer7963 Could you please fix the formatting error by running `pip install -r requirements-dev.txt; ./format.sh`?

WoosukKwon commented on README.md:
```diff
@@ -90,6 +90,7 @@ vLLM seamlessly supports many Hugging Face models, including the following archi
 - StableLM(`stabilityai/stablelm-3b-4e1t`, `stabilityai/stablelm-base-alpha-7b-v2`, etc.)
 - Starcoder2(`bigcode/starcoder2-3b`, `bigcode/starcoder2-7b`, `bigcode/starcoder2-15b`, etc.)
 - Yi (`01-ai/Yi-6B`, `01-ai/Yi-34B`, etc.)
+- Xverse (`xverse/XVERSE-7B-Chat`, `xverse/XVERSE-13B-Chat`, `xverse/XVERSE-65B-Chat`, etc.)
```
nit: Let's keep the alphabetic order :)
```suggestion
- Xverse (`xverse/XVERSE-7B-Chat`, `xverse/XVERSE-13B-Chat`, `xverse/XVERSE-65B-Chat`, etc.)
- Yi (`01-ai/Yi-6B`, `01-ai/Yi-34B`, etc.)
```

WoosukKwon commented on vllm/model_executor/models/xverse.py:
```diff
@@ -0,0 +1,375 @@
+                                              hf_model_weights_iterator)
+from vllm.sequence import SamplerOutput
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]
```
Please replace `KVCache` with `torch.Tensor`. We changed the type recently.

hxer7963 replied:
Thank you for WoosukKwon's reply and suggestions. The issues mentioned in the review have been fixed in the latest commit.

esmeetu commented:
Can we simply use this: `"XverseForCausalLM": ("llama", "LlamaForCausalLM")`?

hxer7963 committed:
Merge branch 'vllm-project:main' into main

hxer7963 committed:
- Fix formatting error with format.sh
- Fix typo in README to keep the alphabetic order
- Replace KVCache with torch.Tensor in xverse.py

hxer7963 commented:
> Can we simply use this: `"XverseForCausalLM": ("llama", "LlamaForCausalLM")`? Will the new moe model use the same name with `XverseForCausalLM`?
Yes, the new moe model will use the same name with `XverseForCausalLM`, so we want to add a new name.

hxer7963 commented:
> Can we simply use this: `"XverseForCausalLM": ("llama", "LlamaForCausalLM")`? Will the new moe model use the same name with `XverseForCausalLM`?

Yes, the new moe model will use the same name with XverseForCausalLM, so we want to add a new name.

hxer7963 commented:
Hello @WoosukKwon ,

I've addressed all the feedback in the PR, and all checks have passed. 

Could you please take another look at the code changes and provide your review?

Thank you for your time and assistance.

Best regards,

hxer7963.

WoosukKwon reviewed (approved):
LGTM! Thanks for submitting the PR. Looking forward to the next model release!
